### PR TITLE
Add boolean dispose flag in HttpRequestStreamReader/ResponseWriter (#540)

### LINKED ISSUE #540: Don't make things null in Dispose
src/Microsoft.AspNet.WebUtilities/HttpRequestStreamReader.Dispose nulls out properties. According to @Tratcher https://github.com/aspnet/HttpAbstractions/pull/539#discussion_r49917289 we should instead set up a disposed bool so we don't accidentally get nullrefs.

### PR DESCRIPTION
**Pull request by jkotalik**
Remove nulling out fields in favor of a dispose boolean.
Fixes #540

### CODE REVIEW

File: src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs [hunk 1]
```diff
@@ -107,23 +108,20 @@ public HttpRequestStreamReader(Stream stream, Encoding encoding, int bufferSize)
                 }
             }
         }
-
+
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs]: (shameless plug) - install this: https://visualstudiogallery.msdn.microsoft.com/bcb3ef8e-a3a4-401e-9e63-fdf00b45268f/view/Reviews and you won't make this mistake again.

**Comment by Tratcher** [comment]
Titles need a description.

**Comment by Tratcher** [comment]
let's do the response stream as well, and check if any other near by components have the same problem.

**Comment by jkotalik** [comment]
🆙  📅  based on what @Tratcher mentioned. Also added tests and modified both files to be consistent.

File: src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs [hunk 2]
```diff
@@ -105,25 +85,23 @@ public HttpRequestStreamReader(Stream stream, Encoding encoding, int bufferSize)
                     _charPool.Return(_charBuffer);
                     _charBuffer = null;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs]: remove these nulls as well.

File: src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs [hunk 3]
```diff
@@ -105,25 +85,23 @@ public HttpRequestStreamReader(Stream stream, Encoding encoding, int bufferSize)
...
 
         protected override void Dispose(bool disposing)
         {
-            if (disposing && _stream != null)
+            if (disposing && !_disposed)
             {
-                _stream = null;
-
+                _disposed = true;
                 if (_bytePool != null)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs]: why are there null checks on _bytePool and _charPool here? The constructor prevents nulls.

File: test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs [hunk 4]
```diff
@@ -191,6 +195,43 @@ public static void ReadLine_ReadWithNoNewlines()
...
+        private static IEnumerable<object[]> HttpRequestStreamReaderData()
+        {
+            yield return new object[] { null, Encoding.UTF8, ArrayPool<byte>.Shared, ArrayPool<char>.Shared };
+            yield return new object[] { new MemoryStream(), null, ArrayPool<byte>.Shared, ArrayPool<char>.Shared };
+            yield return new object[] { new MemoryStream(), Encoding.UTF8, null, ArrayPool<char>.Shared };
+            yield return new object[] { new MemoryStream(), Encoding.UTF8, ArrayPool<byte>.Shared, null };
+        }
+
+        [Fact]
+        public static void NegativeBufferSize_ExpectArgumentOutOfRangeException()
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs]: zero

File: src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs (hunk 3 above)
Comment (jkotalik) [inline src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs]: I think they were from before when we deliberately null out these streams.

Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs]: Except _bytePool wasn't being nulled out before.

File: src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs [hunk 5]
```diff
@@ -132,9 +110,9 @@ protected override void Dispose(bool disposing)
 
         public override int Peek()
         {
-            if (_stream == null)
+            if (_disposed)
             {
-                throw new ObjectDisposedException("stream");
+                throw new ObjectDisposedException(nameof(HttpRequestStreamReader));
```
Comment (Eilon) [inline src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs]: I wouldn't pass anything into the ctor here. The type name is not the "object name" - it's pretty useless here and generates noise.

File: test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs [hunk 6]
```diff
@@ -191,6 +195,43 @@ public static void ReadLine_ReadWithNoNewlines()
             Assert.Null(eol);
         }
 
+        [Theory, MemberData(nameof(HttpRequestStreamReaderData))]
```
Comment (Eilon) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs]: Consider putting the attributes on separate lines.

File: src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs (hunk 5 above)
Comment (jkotalik) [inline src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs]: ObjectDisposedException doesn't actually have a ctor with no arguments. We seem to do the nameof(classname) across our repos.

Comment (Eilon) [inline src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs]: Eh, that's fine. I looked through the CoreFX repo and in there it's a mix of `null`, `GetType().Name`, and other stuff like that. So `nameof(this_class_name)` seems fine.

**Comment by Eilon** [comment]
Would it make sense to add tests that check for `ObjectDisposedException`? If that's a "feature" of this type, should we not test it?

**Comment by jkotalik** [comment]
Sounds reasonable, I'll add some in.

**Comment by jkotalik** [comment]
@Tratcher @Eilon 🆙 📅 . I don't believe the async version can be tested in Assert.Throws, but please correct me if I'm wrong.

**Comment by Tratcher** [comment]
There's an Assert.ThrowsAsync

File: test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs [hunk 7]
```diff
@@ -242,6 +242,17 @@ public static void StreamDisposed_ExpectedObjectDisposedException(Action<HttpReq
             });
         }
 
+        [Fact]
+        public static void StreamDisposed_ExpectObjectDisposedExceptionAsync()
+        {
+            var httpRequestStreamReader = new HttpRequestStreamReader(new MemoryStream(), Encoding.UTF8, 10, ArrayPool<byte>.Shared, ArrayPool<char>.Shared);
+            httpRequestStreamReader.Dispose();
+
+            Assert.ThrowsAsync<ObjectDisposedException>(async () =>
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs]: await this (and make the method public async Task

Comment (jkotalik) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs]: The functions below are already async. I don't understand the value in doing this.

Comment (jkotalik) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs]: `yield return new object[] { new Func<HttpResponseStreamWriter, Task>(async (httpResponseStreamWriter) =>
            {
                await httpResponseStreamWriter.WriteAsync('a');
            })};`

File: test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs [hunk 8]
```diff
@@ -191,6 +195,64 @@ public static void ReadLine_ReadWithNoNewlines()
             Assert.Null(eol);
         }
 
+        [Theory]
+        [MemberData(nameof(HttpRequestNullData))]
+        public static void NullInputsInConstructor_ExpectArgumentNullException(Stream stream, Encoding encoding, ArrayPool<byte> bytePool, ArrayPool<char> charPool)
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs]: Why are your test methods static?

File: test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs [hunk 9]
```diff
@@ -191,6 +195,64 @@ public static void ReadLine_ReadWithNoNewlines()
...
+            httpRequestStreamReader.Dispose();
+
+            Assert.Throws<ObjectDisposedException>(() =>
+            {
+                action(httpRequestStreamReader);
+            });
+        }
+
+        [Fact]
+        public static async void StreamDisposed_ExpectObjectDisposedExceptionAsync()
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs]: async Task, never async void.

File: test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs (hunk 8 above)
Comment (jkotalik) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs]: All of them were static to begin with, should I change it in another PR?

Comment (Tratcher) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs]: you can leave it for now, it's just odd.

Comment (Eilon) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/HttpRequestStreamReaderTest.cs]: `static` means the test class's ctor won't run. Of course, if there isn't one, that's no problem. As long as the test *runs* and it *passes*, it's probably good 😄

**Comment by jkotalik** [comment]
@Eilon this good to go 😄

**Pull request merged**